src/features/device-logs/lib/store.ts

```diff
@@ -48,68 +47,72 @@ const { api } = sbvrUtils;
 
 const supervisor = new Supervisor();
 
-const getWriteContext = async (req: Request): Promise<LogContext> => {
-	const { uuid } = req.params;
-	return await sbvrUtils.db.readTransaction(async (tx) => {
-		const resinApi = api.resin.clone({ passthrough: { req, tx } });
-		const device = (await resinApi.get({
-			resource: 'device',
-			id: { uuid },
-			options: {
-				$select: ['id', 'belongs_to__application'],
-			},
-		})) as PickDeferred<Device, 'id' | 'belongs_to__application'> | undefined;
-		if (!device) {
-			throw new NotFoundError('No device with uuid ' + uuid);
-		}
-		await checkDeviceLogsWritePermissions(device, req, tx);
-		return addRetentionLimit<LogContext>({
-			id: device.id,
-			belongs_to__application: device.belongs_to__application!.__id,
-			uuid,
-		});
-	});
-};
-
-const checkDeviceLogsWritePermissions = (() => {
+const getWriteContext = (() => {
 	const authQuery = _.once(() =>
 		api.resin.prepare<{ id: number }>({
 			method: 'POST',
 			url: `device(@id)/canAccess`,
 			body: { action: 'write-log' },
 		}),
 	);
-	const hasWritePermissions = multiCacheMemoizee(
+	const hasDeviceLogsWritePermissions = async (
+		{ id }: { id: number },
+		req: permissions.PermissionReq,
+		tx: Tx,
+	) => {
+		try {
+			await authQuery()({ id }, undefined, { req, tx });
+			return true;
+		} catch {
+			return false;
+		}
+	};
+	const $getWriteContext = multiCacheMemoizee(
 		async (
-			id: number,
+			uuid: string,
 			req: permissions.PermissionReq,
-			tx: Tx,
-		): Promise<boolean> => {
-			try {
-				await authQuery()({ id }, undefined, { req, tx });
-				return true;
-			} catch {
-				return false;
-			}
+		): Promise<false | LogContext> => {
+			return await sbvrUtils.db.readTransaction(async (tx) => {
+				const resinApi = api.resin.clone({ passthrough: { req, tx } });
+				const device = (await resinApi.get({
+					resource: 'device',
+					id: { uuid },
+					options: {
+						$select: ['id', 'belongs_to__application'],
+					},
+				})) as
+					| PickDeferred<Device, 'id' | 'belongs_to__application'>
+					| undefined;
+				if (!device) {
+					return false;
+				}
+				if (!(await hasDeviceLogsWritePermissions(device, req, tx))) {
+					return false;
```

Comment:
We return false for these and throw the error outside the cache because thrown errors are not cached and we want these failures to be cached (as they're expected errors/behavior rather than exceptions)

Reply:
This makes sense, although I think merging missing Permissions and device not found into `UnauthorizedError()` will make debugging later more tricky.